src/hb-subset-cff1.cc

```diff
@@ -883,186 +701,204 @@ struct cff_subset_plan {
+	if (unlikely (dest == nullptr)) return false;
+	c->push ();
+	if (likely (dest && dest->serialize (c, plan.subset_localsubrs[i])))
+	  subrs_link = c->pop_pack ();
+	else
+	{
+	  c->pop_discard ();
+	  return false;
+	}
+      }
 
-  /* header */
-  cff->version.major = 0x01;
-  cff->version.minor = 0x00;
-  cff->nameIndex = cff->min_size;
-  cff->offSize = 4; /* unused? */
+      PrivateDict  *pd = c->start_embed<PrivateDict> ();
+      if (unlikely (pd == nullptr)) return false;
+      c->push ();
+      cff_private_dict_op_serializer_t privSzr (plan.desubroutinize, plan.drop_hints);
+      /* N.B. local subrs immediately follows its corresponding private dict. i.e., subr offset == private dict size */
+      if (likely (pd->serialize (c, acc.privateDicts[i], privSzr, subrs_link)))
+      {
+	unsigned fd = plan.fdmap[i];
+	plan.fontdicts_mod[fd].privateDictInfo.size = c->length ();
+	plan.fontdicts_mod[fd].privateDictInfo.link = c->pop_pack ();
+      }
+      else
+      {
+	c->pop_discard ();
+	return false;
+      }
+    }
+  }
 
-  /* name INDEX */
+  if (!acc.is_CID ())
+    plan.info.privateDictInfo = plan.fontdicts_mod[0].privateDictInfo;
+
+  /* CharStrings */
   {
-    assert (cff->nameIndex == (unsigned) (c->head - c->start));
-    CFF1NameIndex *dest = c->start_embed<CFF1NameIndex> ();
-    if (unlikely (dest == nullptr)) return false;
-    if (unlikely (!dest->serialize (c, *acc.nameIndex)))
+    CFF1CharStrings  *cs = c->start_embed<CFF1CharStrings> ();
+    if (unlikely (cs == nullptr)) return false;
+    c->push ();
+    if (likely (cs->serialize (c, plan.subset_charstrings)))
+      plan.info.char_strings_link = c->pop_pack ();
+    else
     {
-      DEBUG_MSG (SUBSET, nullptr, "failed to serialize CFF name INDEX");
+      c->pop_discard ();
       return false;
     }
   }
 
-  /* top dict INDEX */
+  /* FDArray (FD Index) */
+  if (acc.fdArray != &Null(CFF1FDArray))
   {
-    assert (plan.offsets.topDictInfo.offset == (unsigned) (c->head - c->start));
-    CFF1IndexOf<TopDict> *dest = c->start_embed< CFF1IndexOf<TopDict>> ();
-    if (dest == nullptr) return false;
-    cff1_top_dict_op_serializer_t topSzr;
-    top_dict_modifiers_t  modifier (plan.offsets, plan.topDictModSIDs);
-    if (unlikely (!dest->serialize (c, plan.offsets.topDictInfo.offSize,
-				    &plan.topdict_mod, 1,
-				    plan.topdict_sizes, topSzr, modifier)))
+    CFF1FDArray  *fda = c->start_embed<CFF1FDArray> ();
+    if (unlikely (fda == nullptr)) return false;
+    c->push ();
+    cff1_font_dict_op_serializer_t  fontSzr;
+    auto it = + hb_zip (+ hb_iter (plan.fontdicts_mod), + hb_iter (plan.fontdicts_mod));
+    if (likely (fda->serialize (c, it, fontSzr)))
+      plan.info.fd_array_link = c->pop_pack (false);
+    else
     {
-      DEBUG_MSG (SUBSET, nullptr, "failed to serialize CFF top dict");
+      c->pop_discard ();
       return false;
     }
   }
 
-  /* String INDEX */
+  /* FDSelect */
+  if (acc.fdSelect != &Null(CFF1FDSelect))
```

Comment:
ditto

Reply:
Unlike FDArray, FDSelect is not an array, has no count field.

Reply:
Meta discussion, seeing places that we have a comparison to Null without no choice making me believe we need something akin to std::optional also maybe.

Reply:
Or just using nullptr instead null pool.

So guess no other change is needed here, let's merge and improve them later.

Thanks :)